src/FunctionApps/SilverTransforms/RecordLinkage/linkage.py

```diff
@@ -0,0 +1,124 @@
+    df.loc[8, 'id'] = df.loc[8, 'id'][:-1] + '3'
+
+    # Near-match case: not actually a dupe, but off by one char
+    df.loc[9, :] = df.loc[2, :]
+    dob_list = df.loc[9, 'birthDate'].split('-')
+    dob_list[1] = str((int(dob_list[1]) + 1) % 12)
+    df.loc[9, 'birthDate'] = '-'.join(dob_list)
+    df.loc[9, 'id'] = df.loc[9, 'id'][:-1] + '8'
+
+    return df[:10]
+
+
+def create_linking_identifier(df: pd.DataFrame):
+    '''
+    Creates a basic identifier used to check duplicates in an incoming
+    batch of data. The basic identifier for the prototype is
+      FIRSTNAME-LASTNAME-YYYY-MM-DD
+    This identifier is joined to the dataframe proper for use in future
+    computation, as well as potential future use in generating/referencing
+    a master ID.
+    '''
+    df['last_name'] = df['name'].apply(lambda x: x[0]['family'] if type(x[0]['family']) is str else '-'.join(x[0]['family']))
+    df['first_name'] = df['name'].apply(lambda x: x[0]['given'] if type(x[0]['given']) is str else '-'.join(list(x[0]['given'])))
+    df['linking_identifier'] = df['first_name'] + '-' + df['last_name'] + '-' + df['birthDate']
+    df.drop(columns=['first_name', 'last_name'], inplace=True)
+    return df
+
+
+def find_matches(df: pd.DataFrame, dupe_field: str):
+    '''
+    From a dataframe containing a column to use as the unique identifier
+    field, identify any and all duplicate records using strict, EXACT
+    match criteria. Returns a dictionary whose keys are the ids of 
+    patients found to be duplicates of some record, and whose values
+    are the new ids to store in the linked resources of these duplicate
+    patients.
+    '''
+    dupes = {}
+    for _, group in df.groupby([dupe_field]):
+        grouped_ids = list(group['id'])
+        dupes[grouped_ids[0]] = list(grouped_ids[1:])
+    ids_to_replace = {}
+    for pid in dupes:
+        if len(dupes[pid]) != 0:
+            for oid in dupes[pid]:
+                ids_to_replace[oid] = pid
+    return ids_to_replace
+
+
+def record_combination_func(x: pd.Series):
+    '''
+    Aggregation function applied behind the scenes when performing
+    de-duplicating linkage. Automatically filters for blank, null,
+    and NaN values to facilitate squashing duplicates down into one
+    consolidated record, such that, for any column X:
+      - if records 1, ..., n-1 are empty in X and record n is not,
+        then n(X) is used as a single value
+      - if some number of records 1, ..., j <= n have the same value 
+        in X and all other records are blank in X, then the value
+        of the matching columns 1, ..., j is used as a singleton
+      - if some number of non-empty in X records 1, ..., j <= n have 
+        different values in X, then all values 1(X), ..., j(X) are
+        concatenated into a unique list of values delimited by |
+    '''
+    if x.isnull().values.all():
+        return ''
+    non_nans = [x for x in list(x.astype(str)) if x != '']
```

Comment:
I'm wondering if `x.astype(str).tolist()` would be safed to ensure we don't get something like `["m", "a", "l", "e", ...]`